README.md
# promiselite

[![Build Status](https://travis-ci.org/D-Mobilelab/promiselite.svg?branch=master&v=1)](https://travis-ci.org/D-Mobilelab/promiselite)
[![Coverage Status](https://coveralls.io/repos/github/D-Mobilelab/promiselite/badge.svg?branch=master&v=1)](https://coveralls.io/github/D-Mobilelab/promiselite?branch=master)
[![npm version](https://badge.fury.io/js/promiselite.svg)](https://badge.fury.io/js/promiselite)
[![Bower version](https://badge.fury.io/bo/promiselite.svg)](https://badge.fury.io/bo/promiselite)
[![GitHub version](https://badge.fury.io/gh/D-Mobilelab%2Fpromiselite.svg?v=1)](https://badge.fury.io/gh/D-Mobilelab%2Fpromiselite)

*PromiseLite* is a light, browser-friendly implementation of JavaScript promises. 

## Installation

### NPM
```
npm install --save promiselite
```
You can found the library ready for production on <i>node_modules/promiselite/dist/dist.js</i>

### Bower
```
bower install --save promiselite
```
You can found the library ready for production on <i>bower_components/promiselite/dist/dist.js</i>

## Documentation

To read documentation, go to:

[http://d-mobilelab.github.io/promiselite/1.3.0](http://d-mobilelab.github.io/promiselite/1.3.0)

Replace <i>1.3.0</i> with the version of the documentation you want to read.


## Usage

### Executors, then and fail

See the documentation of JavaScript *Promise* [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

The name *fail* was used instead of *catch* for compatibility reasons. since some browsers do not allow bare keywords in dot notation.

**Example**

```
#!JavaScript

var p = new PromiseLite(function(resolve, reject){
    setTimeout(function(){
        if (Math.random() > 0.5){
            resolve('It worked!');
        } else {
            reject('It failed!');
        }
    });
}).then(function(value){
    console.info('Yay! ' + value); // outputs 'Yay! It worked!'
}).fail(function(reason){
    console.error('Ops! ' + reason); // outputs 'Ops! It failed!'
});;
```

### Forced blocks

The method *force* corresponds to a *finally* block applied to a chain of promises. It is executed both when a *PromiseLite* instance is both in a fulfilled or rejected state. 

The name *force* was used instead of *finally* for compatibility reasons since some browsers do not allow bare keywords in dot notation.

**Example**

```
#!javascript

var p = new PromiseLite();

p.then(function(value){
   // do something if resolved
   // i.e. read data from an open file
}).fail(function(reason){
   // do something if rejected (or if an Error is raised)
   // i.e. show an error message on screen - "missing file"
}).force(function(){
   // do something both if resolved or rejected
   // i.e. close the file
});
```

### Resolving a Promise created without an executor


```
#!JavaScript

var p = new PromiseLite();

p.then(function(value){
    console.log('The answer is ' + value);
});

// since p has no executor, it won't be resolved until we invoke p.resolve

p.resolve(42); // outputs 'The answer is 42'

```


### PromiseLite.all 


*PromiseLite.all* takes an *Array* of *PromiseLite* as argument and is resolved if and only if all such promises are fulfilled.


```
#!javascript

var p1 = new PromiseLite();
var p2 = new PromiseLite();
var p3 = new PromiseLite();

var pAll = new PromiseLite.all([p1, p2, p3]);

```



### PromiseLite.race


*PromiseLite.race* takes an Array of *PromiseLite* as argument, then

* if the first promise that is settled is fulfilled, the promise returned by *PromiseLite.race* is resolved

* if the first promise that is settled is rejected, the promise returned by *PromiseLite.race* is rejected


```
#!javascript

var p1 = new PromiseLite();
var p2 = new PromiseLite();
var p3 = new PromiseLite();

var pRace = new PromiseLite.race([p1, p2, p3]);

```


### PromiseLite.any



*PromiseLite.any* takes an Array of *PromiseLite* as argument, then

* if at least one of such promises is fulfilled, the promise returned by *PromiseLite.any* is resolved

* if all the promises are rejected, the promise returned by *PromiseLite.any* is rejected


```
#!javascript

var p1 = new PromiseLite();
var p2 = new PromiseLite();
var p3 = new PromiseLite();

var pAny = new PromiseLite.any([p1, p2, p3]);

```


CHANGELOG
** 1.3.2 **
- Fixed promise queue not flushed

** 1.3.1 **
- require is just promise lite

** 1.3.0 **
- Integrated Grunt, travis and coveralls



test/test.js
var expect = function(condition, description){
    if (condition){
        console.info(description);
    } else {
        console.error(description);
    }
}

var it = function(description, func){
    func();
}

var xit = function(description, func){
    // statte xit
    console.log('[IGNORED] - ' + description)
}

it('synchronously resolved promises work', function(){

    var p = new BarneyPromise(function(resolve, reject){
        resolve(47);
    });

    var passedValue;
    p.then(function(data){
        passedValue = data;
    });

    expect(passedValue == 47, 'then passes the value of the promise after resolve');

    var thenConcatValue;

    var thenPromise = p.then(function(data){
        return data + 1;
    }).then(function(data){
        thenConcatValue = data + 1;
    });

    expect(thenConcatValue == 49, 'concatenated then correctly pass data after resolve');
});


it('should catch exceptions in then', function(){

    var pAsync = new BarneyPromise(function(resolve, reject){
        setTimeout(function(){
            resolve({test: 666});
        }, 10);
    }).then(function(obj){
        expect(obj.test === 666, 'async then are managed correctly');
        throw 'testError';
    }, function(reason){
        expect(reason === 'testError', 'catch after then ok');
    });

});

it('should catch errors after reject in executor', function(){
    var pAsync1 = new BarneyPromise(function(resolve, reject){
        setTimeout(function(){
            reject('errorTest');
        }, 100);
    }).then(function(obj){
        expect(true == false, 'should never enter here');
    }, function(reason){
        expect(reason === 'errorTest', 'catch after reject ok');
    });
});

it('should handle chains of catches correctly', function(){

    var pAsync2 = new BarneyPromise(function(resolve, reject){
        setTimeout(function(){
            reject('errorTest');
        }, 100);
    })
    .then(function(obj){
        expect(true == false, 'should never enter here');
    }, function(reason){
        expect(reason === 'errorTest', 'catch (as second argument of then) after reject ok');
        return reason + '1';
    })
    .then(function(obj){
        expect(true == false, 'should never enter here');
    })
    .fail(function(reason){
        expect(pAsync2.isRejected(), 'promise is rejected ');
        expect(reason === 'errorTest1', 'fail after chain ok');
    });

});

it('should not fail if an exception is thrown before the last catch', function(){
    var pAsync3 = new BarneyPromise(function(resolve, reject){
        setTimeout(function(){
            resolve(0);
        }, 100);
    })
    .then(function(num){
        throw 'testThrow';
    }, function(reason){
        expect(reason === 'testThrow', 'catch (as second argument of then) after throw in then');
        return reason + 'Catched';
    })
    .then(function(obj){
        expect(true == false, 'should never enter here');
    })
    .fail(function(reason){
        console.log(reason);
        expect(reason === 'testThrowCatched', 'fail after chain ok');
    });

});

it('should raise exception if there are no catches', function(){
    var pAsync4 = new BarneyPromise(function(resolve, reject){
        setTimeout(function(){
            resolve(0);
        }, 100);
    })
    .then(function(num){
        throw 'throwTest';
    })
    .fail(function(reason){
        console.log(reason);
    });

});


test/tests.js
var runTests = function(PromiseLite, who){

	describe(who + ': Promises without executor', function(){

		it(who + ': should be resolved manually', function(){
			var p = new PromiseLite();

			expect(p.isSettled()).toBe(false);

			p.resolve();

			expect(p.isSettled()).toBe(true);
		});

		it(who + ': should execute then-chains added before they\'re resolved', function(){
			var p = new PromiseLite();

			var thenValue = 42;

			p.then(function(value){
				return value + 1;
			}).then(function(value){
				thenValue = value + 1;
			});

			expect(thenValue).toBe(42);

			p.resolve(0);

			expect(thenValue).toBe(2);
		});

		it(who + ': should execute then-chains added after they\'re resolved', function(){
			var p = new PromiseLite();

			var thenValue = 42;

			expect(thenValue).toBe(42);

			p.resolve(0);

			p.then(function(value){
				return value + 1;
			}).then(function(value){
				return value + 1;
			}).then(function(value){
				thenValue = value + 1;
			});

			expect(thenValue).toBe(3);
		});

		it(who + ': should execute fail if then raises an error', function(){
			var p = new PromiseLite();
			
			var testReason;

			p.then(function(value){
				throw 'testError';
			}).fail(function(reason){
				testReason = reason;
			});

			expect(testReason).toBeUndefined();

			p.resolve();

			expect(testReason).toEqual('testError');
		});

		it(who + ': should throw an exception if then raises an error and there is no fail to catch it', function(){
			var p = new PromiseLite();
			
			var testError;

			p.then(function(value){
				throw 'testError';
			});

			try {
				p.resolve();
			} catch (err){
				testError = err;
			}

			expect(testError).toEqual('testError');
		});

		it(who + ': should throw an exception if fail raises an error and there is no other fail to catch it', function(){
			var p = new PromiseLite();
			
			var testError;

			p.fail(function(reason){
				throw 'testError';
			});

			try {
				p.reject();
			} catch (err){
				testError = err;
			}
			
			expect(testError).toEqual('testError');
		});

		it(who + ': should execute correctly then/fail-chains', function(){
			var p = new PromiseLite();
			
			var thenCalled1, thenCalled2, failCalled1, failCalled2;

			p.fail(function(reason){
				expect(undefined).toBe(null); // this should not be executed
			}).then(function(value){
				thenCalled1 = true;
				expect(value).toEqual(1);
				return value + 1;
			}).then(function(value){
				thenCalled2 = true;
				expect(value).toEqual(2);
				throw 'testError';
			}).then(function(value){
				expect(true).toBe(false); // this should not be executed
			}).fail(function(reason){
				failCalled1 = true;
				expect(reason).toEqual('testError');
				throw 42;
			}).then(function(){
				expect(true).toBe(6); // this should not be called
			}).fail(function(reason){
				failCalled2 = true;
				expect(reason).toEqual(42);
			});

			p.resolve(1);

			expect(thenCalled1).toBe(true);
			expect(thenCalled2).toBe(true);
			expect(failCalled1).toBe(true);
			expect(failCalled2).toBe(true);

		});

	});

	describe(who + ': Promises with executor', function(){

		it(who + ': should be resolved automatically if resolve is invoked in the executor (synchronous function)', function(){
			var p = new PromiseLite(function(resolve, reject){
				resolve();
			});

			expect(p.isFulfilled()).toBe(true);
		});

		it(who + ': should be resolved automatically if resolve is invoked in the executor (asynchronous function)', function(done){
			var p = new PromiseLite(function(resolve, reject){
				setTimeout(resolve, 1000);
			});

			p.then(function(){
				expect(p.isFulfilled()).toBe(true);
				done();
			})
			
		});

		it(who + ': should be rejected automatically if reject is invoked in the executor (synchronous function)', function(){
			var p = new PromiseLite(function(resolve, reject){
				reject();
			});

			expect(p.isRejected()).toBe(true);
		});

		it(who + ': should be rejected automatically if reject is invoked in the executor (asynchronous function)', function(done){
			var p = new PromiseLite(function(resolve, reject){
				setTimeout(reject, 1000);
			});

			p.fail(function(){
				expect(p.isRejected()).toBe(true);
				done();
			})
			
		});

		it(who + ': should throw an exception if the executor raises an error and there is no fail to catch it', function(){
			
			var testError;

			try {
				var p = new PromiseLite(function(resolve, reject){
					throw 'testError';
				});
			} catch (err){
				testError = err;
			}
			
			expect(testError).toEqual('testError');
		});
	});

	describe(who + ': Promise.all', function(){

		it(who + ': should be resolved if all the promises are resolved', function(){
			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAll = new PromiseLite.all([p1, p2, p3]);

			expect(pAll.isPending()).toBe(true);
			
			p1.resolve(1);

			expect(pAll.isPending()).toBe(true);

			p2.resolve(2);

			expect(pAll.isPending()).toBe(true);

			p3.resolve(3);

			expect(pAll.isFulfilled()).toBe(true);

			pAll.then(function(values){
				expect(values[0]).toEqual(1);
				expect(values[1]).toEqual(2);
				expect(values[2]).toEqual(3);
			});

		});

		it(who + ': should be rejected if any of the promises are rejected', function(){
			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAll = new PromiseLite.all([p1, p2, p3]);

			expect(pAll.isPending()).toBe(true);
			
			p1.resolve();

			expect(pAll.isPending()).toBe(true);

			p2.resolve();

			expect(pAll.isPending()).toBe(true);

			p3.reject(3);

			expect(pAll.isRejected()).toBe(true);
			pAll.fail(function(reasons){
				expect(reasons[0]).toBeUndefined();
				expect(reasons[1]).toBeUndefined();
				expect(reasons[2]).toEqual(3);
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAll = new PromiseLite.all([p1, p2, p3]);

			expect(pAll.isPending()).toBe(true);
			
			p1.resolve();

			expect(pAll.isPending()).toBe(true);

			p2.reject(2);

			expect(pAll.isRejected()).toBe(true);
			pAll.fail(function(reasons){
				expect(reasons[0]).toBeUndefined();
				expect(reasons[2]).toBeUndefined();
				expect(reasons[1]).toEqual(2);
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAll = new PromiseLite.all([p1, p2, p3]);

			expect(pAll.isPending()).toBe(true);
			
			p1.reject(3);

			expect(pAll.isRejected()).toBe(true);
			pAll.fail(function(reasons){
				expect(reasons[2]).toBeUndefined();
				expect(reasons[1]).toBeUndefined();
				expect(reasons[0]).toEqual(3);
			});
		});
	});

	describe(who + ': Promise.race', function(){

		it(who + ': should be resolved if the first promise settled is fulfilled', function(){
			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pRace = new PromiseLite.race([p1, p2, p3]);

			expect(pRace.isPending()).toBe(true);
			
			p2.resolve(2);

			expect(pRace.isFulfilled()).toBe(true);

			pRace.then(function(values){
				expect(values[0]).toBeUndefined();
				expect(values[1]).toEqual(2);
				expect(values[2]).toBeUndefined;
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pRace = new PromiseLite.race([p1, p2, p3]);

			expect(pRace.isPending()).toBe(true);
			
			p1.resolve(2);

			expect(pRace.isFulfilled()).toBe(true);

			pRace.then(function(values){
				expect(values[2]).toBeUndefined();
				expect(values[0]).toEqual(2);
				expect(values[1]).toBeUndefined;
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pRace = new PromiseLite.race([p1, p2, p3]);

			expect(pRace.isPending()).toBe(true);
			
			p3.resolve(2);

			expect(pRace.isFulfilled()).toBe(true);

			pRace.then(function(values){
				expect(values[0]).toBeUndefined();
				expect(values[2]).toEqual(2);
				expect(values[1]).toBeUndefined;
			});

		});

		it(who + ': should be rejected if the first promise settled is rejected', function(){
			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pRace = new PromiseLite.race([p1, p2, p3]);

			expect(pRace.isPending()).toBe(true);
			
			p2.reject(2);

			expect(pRace.isRejected()).toBe(true);

			pRace.fail(function(reasons){
				expect(reasons[0]).toBeUndefined();
				expect(reasons[1]).toEqual(2);
				expect(reasons[2]).toBeUndefined;
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pRace = new PromiseLite.race([p1, p2, p3]);

			expect(pRace.isPending()).toBe(true);
			
			p1.reject(2);

			expect(pRace.isRejected()).toBe(true);

			pRace.fail(function(reasons){
				expect(reasons[1]).toBeUndefined();
				expect(reasons[0]).toEqual(2);
				expect(reasons[2]).toBeUndefined;
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pRace = new PromiseLite.race([p1, p2, p3]);

			expect(pRace.isPending()).toBe(true);
			
			p3.reject(2);

			expect(pRace.isRejected()).toBe(true);

			pRace.fail(function(reasons){
				expect(reasons[0]).toBeUndefined();
				expect(reasons[2]).toEqual(2);
				expect(reasons[1]).toBeUndefined;
			});

		});
		
	});

	describe(who + ': Promise.any', function(){

		it(who + ': should be resolved as soon as one promise is resolved', function(){
			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAny = new PromiseLite.any([p1, p2, p3]);

			expect(pAny.isPending()).toBe(true);
			
			p1.reject();
			expect(pAny.isPending()).toBe(true);

			p2.reject();
			expect(pAny.isPending()).toBe(true);
			
			p3.resolve(2);
			expect(pAny.isFulfilled()).toBe(true);

			pAny.then(function(values){
				expect(values[0]).toBeUndefined();
				expect(values[1]).toBeUndefined();
				expect(values[2]).toEqual(2);
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAny = new PromiseLite.any([p1, p2, p3]);

			expect(pAny.isPending()).toBe(true);
			
			p1.reject();
			expect(pAny.isPending()).toBe(true);

			p2.resolve(1);
			expect(pAny.isFulfilled()).toBe(true);
			
			p3.reject();

			pAny.then(function(values){
				expect(values[0]).toBeUndefined();
				expect(values[1]).toEqual(1);
				expect(values[2]).toBeUndefined();
			});

			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAny = new PromiseLite.any([p1, p2, p3]);

			expect(pAny.isPending()).toBe(true);
			
			p1.resolve(3);
			expect(pAny.isFulfilled()).toBe(true);
			
			p2.reject();

			pAny.then(function(values){
				expect(values[0]).toEqual(3);
				expect(values[1]).toBeUndefined();
				expect(values[2]).toBeUndefined();
			});

		});

		it(who + ': should be rejected if all the promises are rejected', function(){
			
			var p1 = new PromiseLite();
			var p2 = new PromiseLite();
			var p3 = new PromiseLite();

			var pAny = new PromiseLite.any([p1, p2, p3]);

			expect(pAny.isPending()).toBe(true);
			
			p1.reject(1);
			p2.reject(2);
			p3.reject(3);

			pAny.fail(function(reasons){
				expect(reasons[0]).toEqual(1);
				expect(reasons[1]).toEqual(2);
				expect(reasons[2]).toEqual(3);
			});

		});
		
	});

	describe(who + ': Promise.force', function(){

		it(who + ': should be executed when the promise is resolved', function(){
			
			var p1 = new PromiseLite();
			
			p1.force(function(value){
				expect(value).toEqual(1);
			});

			p1.resolve(1);

		});

		it(who + ': should be executed when the promise is rejected', function(){
			
			var p1 = new PromiseLite();
			
			p1.force(function(reason){
				expect(reason).toEqual(1);
			});

			p1.reject(1);

		});

		it(who + ': should be executed when a then raises an error', function(){
			
			var p1 = new PromiseLite();
			
			p1.then(function(){
				throw 'testError';
			}).force(function(value){
				expect(value).toEqual('testError');
			});

			p1.resolve();

		});

		it(who + ': should be executed when a fail raises an error', function(){
			
			var p1 = new PromiseLite();
			
			p1.then(function(){
				throw 'testError'
			}).fail(function(){
				throw 'anotherError';
			}).then(function(){
				expect(true).toBe(false); // should not enter here
			}).force(function(value){
				expect(value).toEqual('anotherError');
			});

			p1.resolve();

		});

		it(who + ': should be executed in chain', function(){
			
			var p1 = new PromiseLite();
			
			p1.force(function(){
				throw 'error1';
			}).fail(function(reason){
				expect(reason).toEqual('error1');
				throw 'error2';
			}).force(function(value){
				expect(value).toEqual('error2');
				throw 'error3';
			}).then(function(value){
				expect(true).toBe(false); // should not enter here
			}).force(function(reason){
				expect(reason).toEqual('error3')
			});

			p1.resolve();

		});
		
	});

describe(who + ': updatable promises', function(){

		it(who + ': should be resolved N times, given N in the constructor', function(){
			
			var resolveCount = 0;

			var p1 = new PromiseLite(undefined, 3);
			
			p1.then(function(){
				resolveCount++;
			});

			p1.resolve();
			expect(resolveCount).toEqual(1);
			p1.resolve();
			expect(resolveCount).toEqual(2);
			p1.resolve();
			expect(resolveCount).toEqual(3);
			p1.resolve();
			expect(resolveCount).toEqual(3);

		});
		
	});

}

module.exports = runTests;

promise.js
var PromiseLite = (function(){

    var PrivatePromise = function(executor, nextProm){

        // executor called at the end of the definition of Promise
        if (typeof executor !== 'undefined' && typeof executor !== 'function'){
            throw 'PromiseLite :: executor must be a function, got ' + typeof executor;
        }
        
        var promiseInstance = this;
        var promiseStatusIndex = 0;
        var promiseValue;
        var promiseReason;
        var next, lastNext;

        if (nextProm){
            next = nextProm;
        }

        var promiseStatus = {
            0: 'pending',
            1: 'fulfilled',
            2: 'rejected'
        }

        var NOOP = function(){};
        var PASS = function(arg){
            return arg;
        }

        var getValue = function(){
            return promiseValue;
        }

        var getReason = function(){
            return promiseReason;
        }

        this.isPending = function(){
            return promiseStatusIndex === 0;
        }

        this.isFulfilled = function(){
            return promiseStatusIndex === 1;
        }

        this.isRejected = function(){
            return promiseStatusIndex === 2;
        }

        this.isSettled = function(){
            return (promiseStatusIndex === 1) || (promiseStatusIndex === 2);
        }

        this.getStatus = function(){
            return promiseStatus[promiseStatusIndex];
        }

        var immediatelyFulfill = function(success, error, next, _getValue){

            if (typeof _getValue === 'undefined'){
                _getValue = getValue;
            }

            if (typeof success === 'undefined'){
                success = NOOP;
            }
            if (typeof error === 'undefined'){
                error = PASS;
            }

            return new PrivatePromise(function(res, rej){
                try {
                    res(success(_getValue()));
                } catch (err){
                    // if we're trying to pass the error to the next node of the chain
                    // but the next node of the chain is undefined
                    // throw error, otherwise pass it forward through the chain
                    if (error == PASS && typeof next === 'undefined'){
                        throw err;
                    } else {
                        rej(error(err));   
                    }
                }
            }, next);

        }

        var immediatelyReject = function(error, next, _getReason){

            if (typeof _getReason === 'undefined'){
                _getReason = getReason;
            }

            if (typeof error === 'undefined'){
                error = PASS;
            }

            return new PrivatePromise(function(res, rej){ 
                rej(error(_getReason()));
            }, next);
            
        }

        this.resolve = function(value){
            if (promiseStatusIndex === 1){
                return;
            }
            promiseStatusIndex = 1;
            promiseValue = value;

            if (next){
                immediatelyFulfill(next.onSuccess, next.onError, next.next)
            }
        }

        this.reject = function(reason){
            if (promiseStatusIndex === 2){
                return;
            }
            promiseStatusIndex = 2;
            promiseReason = reason;

            if (next){
                immediatelyReject(next.onError, next.next)
            }
        }

        var setNext = function(onSuccess, onError){

            if (typeof onError === 'undefined'){
                onError = PASS;
            }

            var createNext = function(){
                return {
                    next: undefined,
                    onSuccess: onSuccess,
                    onError: onError
                }
            }

            if (typeof next === 'undefined'){
                next = createNext();
            } else {
                lastNext = next;        
                while (typeof lastNext.next !== 'undefined'){
                    lastNext = lastNext.next;
                }
                lastNext.next = createNext();
            }
        }

        this.then = function(onSuccess, onError){
            if (promiseInstance.isPending()){
                setNext(onSuccess, onError);
                return promiseInstance;
            }

            if (promiseInstance.isFulfilled()){
                return immediatelyFulfill(onSuccess, onError);
            }

            if (promiseInstance.isRejected()){
                return immediatelyReject(onError);
            }
        }

        this.fail = function(onError){
            return promiseInstance.then(undefined, onError);
        }

        if (typeof executor === 'function'){
            executor(promiseInstance.resolve, promiseInstance.reject);
        }
    
    }
    
    var PublicPromise = function(executor){
        return new PrivatePromise(executor);
    }

    return PublicPromise;

})();

if (typeof module === 'object' && module.exports) {
    module.exports = PromiseLite;
} else {
    window.PromiseLite = PromiseLite;
}

karma.conf.js
// Karma configuration
// http://karma-runner.github.io/0.12/config/configuration-file.html
// Generated on 2015-05-15 using
// generator-karma 1.0.0

module.exports = function(config) {
    'use strict';

    config.set({
        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: false,

        // base path, that will be used to resolve files and exclude
        basePath: '.',

        // testing framework to use (jasmine/mocha/qunit/...)
        // as well as any additional frameworks (requirejs/chai/sinon/...)
        frameworks: [
            "jasmine",
            "browserify",
        ],

        // list of files / patterns to load in the browser
        files: [
            "mock.js",
            // "src/**/*.js",
            "test/**/*.js"
        ],

        // list of files / patterns to exclude
        exclude: [
            "test/coverage/*"
        ],

        // web server port
        port: 8080,

        // Start these browsers, currently available:
        // - Chrome, ChromeCanary, Firefox, Opera, Safari (only Mac), PhantomJS, IE (only Windows)
        browsers: [
            "PhantomJS"
        ],

        // Which plugins to enable
        plugins: [
            "karma-phantomjs-launcher",
            "karma-jasmine",
            "karma-coverage",
            "karma-browserify",
            "karma-coveralls",
            // "karma-mocha-reporter"
        ],

        // Continuous Integration mode
        // if true, it capture browsers, run tests and exit
        singleRun: false,

        colors: true,

        // level of logging
        // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
        logLevel: config.LOG_INFO,

        reporters: [
            'progress', 
            'coverage',
            'coveralls',
            // 'mocha'
        ],
        preprocessors: {
            'src/**/*.js': ['coverage'],
            'test/*.test.js': ['browserify']
        },
        coverageReporter: {
            type : 'lcov',
            dir : 'test/coverage/'
        },

        // enable / disable colors in the output (reporters and logs)
        colors: true,

        browserify: {
            debug: true,
            transform: ['browserify-istanbul']
        }

        // Uncomment the following lines if you are using grunt's server to run the tests
        // proxies: {
        //   '/': 'http://localhost:9000/'
        // },
        // URL root prevent conflicts with the site root
        // urlRoot: '_karma_'
});
};


demo/demo.js
var expect = function(condition, description){
    if (condition){
        console.info(description);
    } else {
        console.error(description);
    }
}

var it = function(description, func){
    func();
}

var xit = function(description, func){
    // statte xit
    console.log('[IGNORED] - ' + description)
}

it('synchronously resolved promises work', function(){

    var p = new PromiseLite(function(resolve, reject){
        resolve(47);
    });

    var passedValue;
    p.then(function(data){
        passedValue = data;
    });

    expect(passedValue == 47, 'then passes the value of the promise after resolve');

    var thenConcatValue;

    var thenPromise = p.then(function(data){
        return data + 1;
    }).then(function(data){
        thenConcatValue = data + 1;
    });

    expect(thenConcatValue == 49, 'concatenated then correctly pass data after resolve');
});


it('should catch exceptions in then', function(){

    var pAsync = new PromiseLite(function(resolve, reject){
        setTimeout(function(){
            resolve({test: 666});
        }, 10);
    }).then(function(obj){
        expect(obj.test === 666, 'async then are managed correctly');
        throw 'testError';
    }, function(reason){
        expect(reason === 'testError', 'catch after then ok');
    });

});

it('should catch errors after reject in executor', function(){
    var pAsync1 = new PromiseLite(function(resolve, reject){
        setTimeout(function(){
            reject('errorTest');
        }, 100);
    }).then(function(obj){
        expect(true == false, 'should never enter here');
    }, function(reason){
        expect(reason === 'errorTest', 'catch after reject ok');
    });
});

it('should handle chains of catches correctly', function(){

    var pAsync2 = new PromiseLite(function(resolve, reject){
        setTimeout(function(){
            reject('errorTest');
        }, 100);
    })
    .then(function(obj){
        expect(true == false, 'should never enter here');
    }, function(reason){
        expect(reason === 'errorTest', 'catch (as second argument of then) after reject ok');
        return reason + '1';
    })
    .then(function(obj){
        expect(true == false, 'should never enter here');
    })
    .fail(function(reason){
        expect(pAsync2.isRejected(), 'promise is rejected ');
        expect(reason === 'errorTest1', 'fail after chain ok');
    });

});

it('should not fail if an exception is thrown before the last catch', function(){
    var pAsync3 = new PromiseLite(function(resolve, reject){
        setTimeout(function(){
            resolve(0);
        }, 100);
    })
    .then(function(num){
        throw 'testThrow';
    }, function(reason){
        expect(reason === 'testThrow', 'catch (as second argument of then) after throw in then');
        return reason + 'Catched';
    })
    .then(function(obj){
        expect(true == false, 'should never enter here');
    })
    .fail(function(reason){
        console.log(reason);
        expect(reason === 'testThrowCatched', 'fail after chain ok');
    });

});

it('should raise exception if there are no catches', function(){
    var pAsync4 = new PromiseLite(function(resolve, reject){
        setTimeout(function(){
            resolve(0);
        }, 100);
    })
    .then(function(num){
        throw 'throwTest';
    })
    .fail(function(reason){
        console.log(reason);
    });

});
